### ISSUE TITLE
maintenance daemon might get stuck waiting for pg_dist_node lock

### ISSUE THREAD

**Issue by marcocitus** [citusdata/citus#3065]
The maintenance daemon currently takes an exclusive lock on pg_dist_node on every iteration:
https://github.com/citusdata/citus/blob/ec9fee1c922cf3690fb2bb5af699c0e045c653a5/src/backend/distributed/metadata/metadata_sync.c#L1281

It can easily get stuck waiting for a large transaction that takes a somewhat aggressive `pg_dist_node` lock such as `create_distributed_function`. If that transaction is in a distributed deadlock, the maintenance daemon is unable to resolve it.
https://github.com/citusdata/citus/blob/29f1ea079b975816b0248788bc9a0d70beaa273f/src/backend/distributed/commands/function.c#L607

We should perhaps add some timeout to the lock and maybe use a less aggressive lock.

**Comment by serprex** [comment]
To reproduce issue, interleave these sessions
```
session 1:
create table t2(a int, b int);select create_distributed_table('t2', 'a');insert into t2 values (1, 2), (2, 3);begin;update pg_dist_node set metadatasynced = metadatasynced;
commit;begin;update pg_dist_node set metadatasynced = metadatasynced;
update t2 set b = 3 where a = 1;
update t2 set b = 3 where a = 2;

session 2:
select start_metadata_sync_to_node('localhost', 5434);
begin;update t2 set b = 4 where a = 2;
update t2 set b = 4 where a = 1;
```

**Comment by pykello** [comment]
> https://github.com/citusdata/citus/blob/29f1ea079b975816b0248788bc9a0d70beaa273f/src/backend/distributed/commands/function.c#L607
> 
> We should perhaps add some timeout to the lock and maybe use a less aggressive lock.

I think the idea was this we wanted to get a lock that allows concurrent reads, but blocks on updates. I am not sure if I understand what we should do when this times out. Just ignore doing this or schedule it for maintenanced? What do you think instead of doing this, always schedule it for maintenanced?

**Comment by pykello** [comment]
> > https://github.com/citusdata/citus/blob/29f1ea079b975816b0248788bc9a0d70beaa273f/src/backend/distributed/commands/function.c#L607
> > 
> > We should perhaps add some timeout to the lock and maybe use a less aggressive lock.
> 
> I think the idea was this we wanted to get a lock that allows concurrent reads, but blocks on updates. I am not sure if I understand what we should do when this times out. Just ignore doing this or schedule it for maintenanced? What do you think instead of doing this, always schedule it for maintenanced?

Actually I think we should postpone this to maintenanced, or otherwise we have to get a lock in `SyncMetadataToNodes()` that conflicts with the lock here. Otherwise race conditions can happen, examples of which is failures we see in https://circleci.com/gh/citusdata/citus/43478?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link.

**Comment by marcocitus** [comment]
> I think the idea was this we wanted to get a lock that allows concurrent reads, but blocks on updates.

I think that's fine. I was mostly giving it as an example of a place where we take a lock that conflicts with the lock taken by the maintenance daemon because it happened to be the one I encountered, but there are others as well. We mainly want to lower the lock level in the maintenance daemon.

**Comment by marcocitus** [comment]
Fixed by #3068
